docs/design/Ratify v2 Architecture Proposal.md

```diff
@@ -0,0 +1,262 @@
+}
+type ReferenceVerifier interface {
+	VerifyReference(store ReferrerStore) bool
+}
+type Executor struct {
+	verifiers map[string]ReferenceVerifier
+	stores    map[string]ReferrerStore
+}
+var verifierTypes map[string]ReferenceVerifier
+var storeTypes map[string]ReferrerStore
+
+func RegisterStore(name string, store ReferrerStore) {
+	fmt.Printf("Registering store %s\n", name)
+	if storeTypes == nil {
+		storeTypes = make(map[string]ReferrerStore)
+	}
+	storeTypes[name] = store
+}
+
+func RegisterVerifier(name string, verifier ReferenceVerifier) {
+	fmt.Printf("Registering verifier %s\n", name)
+	if verifierTypes == nil {
+		verifierTypes = make(map[string]ReferenceVerifier)
+	}
+	verifierTypes[name] = verifier
+}
+
+func (r *Executor) Verify() {
+	r.verifiers["cosign"].VerifyReference(r.stores["oras"])
+}
+
+func NewExecutor() *Executor {
+	return &Executor{
+		verifiers: verifierTypes,
+		stores:    storeTypes,
+	}
+}
+/*----------------------------------------------------------*/
+// notation verifier
+type Verifier struct{}
+
+func (v *Verifier) VerifyReference(store core.ReferrerStore) bool {
+	fmt.Println("Notation verifier")
+	fmt.Println(store.ListReferrers())
+	return true
+}
+
+func init() {
+	core.RegisterVerifier("notation", &Verifier{})
+}
+/*----------------------------------------------------------*/
+// cosign verifier
+type Verifier struct{}
+
+func (v *Verifier) VerifyReference(store core.ReferrerStore) bool {
+	fmt.Println("Cosign verifier")
+	fmt.Println(store.ListReferrers())
+	return true
+}
+
+func init() {
+	core.RegisterVerifier("cosign", &Verifier{})
+}
+
+/*----------------------------------------------------------*/
+// oras store
+type Store struct{}
+
+func (s *Store) ListReferrers() string {
+	fmt.Println("oras")
+	return "oras"
+}
+
+func init() {
+	core.RegisterStore("oras", &Store{})
+}
+
+/*----------------------------------------------------------*/
+// Ratify main application
+package main
+
+import (
+	"github.com/binbin-li/ratify-test/core"
+)
+
+func main() {
+	manager := core.NewExecutor()
+	manager.Verify()
+}
+```
+```Dockerfile
+# Dockerfile
+# Use the official Golang image as the base image
+FROM golang:1.21-alpine
+
+# Set the working directory inside the container
+WORKDIR /app
+
+# Copy the Go modules files
+COPY go.mod ./
+
+# Download the Go modules
+# RUN go mod download
+
+# Copy the source code
+COPY . .
+
+# Add appropriate verifier/store implementation
+RUN sed -i '2i\import _ "github.com/binbin-li/ratify-test/cosign"' main.go && \
+    sed -i '2i\import _ "github.com/binbin-li/ratify-test/notation"' main.go && \
+    sed -i '2i\import _ "github.com/binbin-li/ratify-test/oras"' main.go
+
+RUN go mod tidy
+
+# Build the Go application
+RUN go build -o main .
+
+# Command to run the executable
+CMD ["./main"]
+```
+In the above example, the only dependency of Ratify repo is the Ratify core library. And users could select appropriate interface implementations based on need in the Dockerfile.
```

Comment:
just clarify, we will have docker files for each entry point ( cli , k8s)?

Reply:
Good question!
1. For different repos, like ratify-cli, ratify, they will different scripts and Dockerfile for each. 
2. For ratify repo, users might have different scenarios to integrate different plugins. I plan to have one Dockerfile along with a script so that users can just pass in different plugin paths to it instead of writing different Dockerfiles.
3. And ratify repo should also ship one or a few images with default verifiers that cover common user scenarios.